177 unified sdk staking (#198)
casper-network/docs-new: source/docs/casper/developers/dapps/sdk/client-library-usage.md

</Tabs>

Once submitted, the above snippet will print the deploy hash in the console.

## Staking

> Token staking is vital to the Casper Network, enabling its delegated Proof-of-Stake consensus mechanism to operate effectively. With the help of any of the Casper SDKs, you can stake tokens by delegating them to validators.

Reviewer: ```suggestion
Token staking is a fundamental aspect of the Casper Network, whereby users lock up tokens as collateral in exchange for the ability to participate in the blockchain's consensus mechanism and earn rewards. This delegated Proof-of-Stake consensus mechanism is crucial for the network's effective operation. With the aid of any of the Casper SDKs, you can delegate your tokens to validators and participate in staking on the network.
```

> The delegation functionality exists as a piece of session code obtainable from the [casper-node](https://github.com/casper-network/casper-node) repository. To delegate tokens, first clone the repository:

Reviewer: ```suggestion
The delegation functionality is available as a smart contract, which can be found in the [casper-node](https://github.com/casper-network/casper-node) repository. To delegate tokens, first clone the repository:
```

```bash
git clone https://github.com/casper-network/casper-node.git
cd casper-node/
```

Then compile the [delegate contract](https://github.com/casper-network/casper-node/blob/dev/smart_contracts/contracts/client/delegate/src/main.rs):

```bash
make setup-rs
make build-contract-rs/delegate
```

> Now, assuming that you cloned `casper-node` from your project's root directory, `cd` back into it:
> 
> ```bash
> cd ../

Reviewer: ```suggestion
The complied Wasm can be found in this directory: `target/wasm32-unknown-unknown/release/`.

```bash
ls target/wasm32-unknown-unknown/release/delegate.wasm

Author: This section is not to show the output directory, but to explain that if they cloned from their project's root directory, they can go back to their root directory and access *delegate.wasm* via the path in the code example. This could be altered but not in the way you've suggested IMO.

```

Now in your dApp's backend (or standalone script), load the *delegate.wasm* file into memory and deploy it with the arguments "amount", "delegator", and "validator" included.

<Tabs>

<TabItem value="js" label="JavaScript">

```javascript

> const { CasperClient, Contracts, RuntimeArgs, CLValueBuilder, CLPublicKey } = require("casper-js-sdk")
> const fs = require("fs")

Reviewer: ```suggestion
const { CasperClient, Contracts, RuntimeArgs, CLValueBuilder, CLPublicKey } = require("casper-js-sdk");
const fs = require("fs");
```

> const casperClient = new CasperClient("http://NODE_ADDRESS:7777/rpc")
> const contract = new Contracts.Contract(casperClient)

Reviewer: ```suggestion
const casperClient = new CasperClient("http://NODE_ADDRESS:7777");
const contract = new Contracts.Contract(casperClient);
```

Author: As far as I'm aware "/rpc" is required. Have you tested this and found otherwise?

const contractWasm = new Uint8Array(fs.readFileSync("./casper-node/target/wasm32-unknown-unknown/release/delegate.wasm").buffer)

const runtimeArguments = RuntimeArgs.fromMap({
  "amount": CLValueBuilder.u512(5e11), // 500 CSPR, minimum delegation amount
	"delegator": keypair.publicKey,
  "validator": CLPublicKey.fromHex("01e8c84f4fbb58d37991ef373c08043a45c44cd7f499453fa2bd3e141cc0113b3c")
})

> const deploy = contract.install(
> 	contractWasm,
> 	runtimeArguments,
> 	"5000000000", // Gas payment (5 CSPR)
> 	keypair.publicKey,
> 	"casper", // or "casper-test" for testnet
> 	[keypair]
> )

Reviewer: ```suggestion
const deploy = contract.install(
    contractWasm,
    runtimeArguments,
    "5000000000", // Gas payment (5 CSPR)
    keypair.publicKey,
    "casper", // or "casper-test" for testnet
    [keypair]
);
```

console.log(await casperClient.putDeploy(deploy))
```

</TabItem>

<TabItem value="python" label="Python">

```python
import pycspr

> validator_public_key = pycspr.factory.accounts.create_public_key_from_account_key(
> 	bytes.fromhex("01e8c84f4fbb58d37991ef373c08043a45c44cd7f499453fa2bd3e141cc0113b3c")
> )

Reviewer: ```suggestion
validator_public_key = pycspr.factory.accounts.create_public_key_from_account_key(
    bytes.fromhex("01e8c84f4fbb58d37991ef373c08043a45c44cd7f499453fa2bd3e141cc0113b3c")
)
```

> deploy_params = pycspr.create_deploy_parameters(
> 	account = keypair, # Only the public key is used, see `create_deploy_parameters`
> 	chain_name = "casper" # or "casper-test" for testnet
> )

Reviewer: ```suggestion
deploy_params = pycspr.create_deploy_parameters(
    account = keypair, # Only the public key is used, see `create_deploy_parameters`
    chain_name = "casper" # or "casper-test" for testnet
)
```

> deploy = pycspr.create_validator_delegation(
> 	params = deploy_params,
> 	amount = int(5e11), # 500 CSPR, minimum delegation amount
> 	public_key_of_delegator = keypair,
> 	public_key_of_validator = validator_public_key,
> 	path_to_wasm = "./casper-node/target/wasm32-unknown-unknown/release/delegate.wasm"
> )

Reviewer: ```suggestion
deploy = pycspr.create_validator_delegation(
    params = deploy_params,
    amount = int(500e9), # Minimum delegation amount:  500 CSPR
    public_key_of_delegator = keypair,
    public_key_of_validator = validator_public_key,
    path_to_wasm = "./casper-node/target/wasm32-unknown-unknown/release/delegate.wasm"
)
```
